### ISSUE TITLE
todo

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/improving-transformers-world-model-for-rl#1]
- [ ] world model
  - [ ] do 3d axial rotary embeddings, with more frequencies allotted for time
  - [x] sampling trajectories
  - [x] key / value caching in block causal context
  - [x] make sure block size 1 falls back to regular causal
  - [x] actions + rewards, bring in hl gauss for more generalizable
  - [ ] make sure it works with vector quantize (residual optional) with Fifty's new rotation trick
  - [ ] handle start token, for sampling sans prompting
- [x] ppo
  - [ ] allow for world model to be conditioned on the returns
- [ ] rest of the network, impala cnn + rnn, though not the meat of the paper
- [x] environment
  - [x] figure out if pufferlib is working on craftax bindings, and if not, look into interoperable torch2jax packages
- [ ] improvisations
  - [ ] allow for the state tokens in the world model to be conditioned by the distance to the nearest code id, perhaps it can make use of that uncertainty during training
  - [ ] think about how to involve a discriminator in some GAN like system, given the small chance that GAN stability has been mitigated / solved by a recent paper out of Cornell / Brown
- [x] use tensor typing for this project
- [x] follow up work
  - [x] "we plan to modify the policy to directly accept latent tokens generated by the TWM."
  - [x] replay the world model embed correctly during learning phase, if incoporated
  - [ ] explore turning the world model into a policy network. probably converges on some sort of online decision transformer at that point
    - [ ] add grpo, either with world model as policy network, or policy network connected at the bottom with frozen world model. offer a few ways for them to receive gradients, starting off with gumbel softmax
    - [ ] just quickly try an idea where the world model, actor, critic are all one transformer in [another repo](https://github.com/lucidrains/ppo/blob/master/train_world_model.py)
